### PR TITLE
MNT: improve the error message in Path init

### PR DESCRIPTION
**Pull request by tacaswell**
Improve the error messages when invalidly shaped data is passed to
`matplotlib.path.Path.__init__`.

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant

### CODE REVIEW

File: lib/matplotlib/path.py [hunk 1]
```diff
@@ -128,16 +128,20 @@ def __init__(self, vertices, codes=None, _interpolation_steps=1,
         vertices = _to_unmasked_float_array(vertices)
         if vertices.ndim != 2 or vertices.shape[1] != 2:
             raise ValueError(
-                "'vertices' must be a 2D list or array with shape Nx2")
+                "'vertices' must be a 2D list or array with shape (N, 2). "
+                f"Your input has shape {vertices.shape}.")
```
Comment (anntzer) [inline lib/matplotlib/path.py]: I was thinking of also having a `cbook._check_shape((None, 2), vertices=vertices)` similar to _check_isinstance, _check_in_list, _check_getitem, given that this also comes up quite often...

Comment (timhoffm) [reply lib/matplotlib/path.py]: Sounds reasonable.

Comment (tacaswell) [reply lib/matplotlib/path.py]: Is that a "please do this for this PR" or "this seems like a good idea we should think about"?

**Comment by timhoffm** [comment]
Flake8:

./lib/matplotlib/tests/test_path.py:32:1: E302 expected 2 blank lines, found 1

File: lib/matplotlib/path.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/path.py]: Both ways are ok for me.

Comment (anntzer) [reply lib/matplotlib/path.py]: It's a "let's try and see whether @tacaswell is going to bite the hook and do it here, but I won't block over that either".

Comment (tacaswell) [reply lib/matplotlib/path.py]: :fishing_pole_and_fish:

File: lib/matplotlib/path.py [hunk 2]
```diff
@@ -126,18 +126,19 @@ def __init__(self, vertices, codes=None, _interpolation_steps=1,
...
-            raise ValueError(
-                "'vertices' must be a 2D list or array with shape Nx2")
+        cbook._check_shape((None, 2), vertices=vertices)
 
         if codes is not None:
             codes = np.asarray(codes, self.code_type)
             if codes.ndim != 1 or len(codes) != len(vertices):
                 raise ValueError("'codes' must be a 1D list or array with the "
-                                 "same length of 'vertices'")
+                                 "same length of 'vertices'. "
```
Comment (tacaswell) [inline lib/matplotlib/path.py]: I think the specific text here about verts and codes is worth not re-using the general checking logic.

File: lib/matplotlib/cbook/__init__.py [hunk 3]
```diff
@@ -2194,6 +2194,48 @@ def _check_in_list(_values, **kwargs):
...
+                for t, s in zip(target_shape, data_shape)
+            )
+        ):
+            def format_dims(target_shape):
+                dim_labels = iter(itertools.chain(
+                    'MNLIJKLH',
+                    (f"D{i}" for i in itertools.count())))
+                text_shape = tuple(n if n is not None else next(dim_labels)
+                                   for n in target_shape)
+                return '(' + ", ".join(str(_) for _ in text_shape) + ')'
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: fstrings are great but I don't think this warrants being its own inner function?

File: lib/matplotlib/cbook/__init__.py [hunk 4]
```diff
@@ -2194,6 +2194,48 @@ def _check_in_list(_values, **kwargs):
...
+    for k, v in kwargs.items():
+        data_shape = v.shape
+
+        if not (
+            len(data_shape) == len(target_shape)
+            and all(
+                (t == s if t is not None else True)
+                for t, s in zip(target_shape, data_shape)
+            )
+        ):
```
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]:
```
        if (len(data_shape) != len(target_shape)
                or any(t not in [d, None]
                       for d, t in zip(data_shape, target_shape))):
```
(not so much the formatting (heh :-)), just the inverted logic seems easier to read?)

**Comment by anntzer** [comment]
happy to have successfully nerd-sniped you :p (but that was indeed something in my todo since a while)

File: lib/matplotlib/path.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/path.py]: sounds fine

File: lib/matplotlib/cbook/__init__.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py]: fair

File: lib/matplotlib/cbook/__init__.py (hunk 4 above)
Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py]: I had it that way eventually and thought the `and` was clearer than the `or`.  the `t not in [d, None]` is much nicer than the ternary though.

File: lib/matplotlib/path.py [hunk 5]
```diff
@@ -126,18 +126,19 @@ def __init__(self, vertices, codes=None, _interpolation_steps=1,
...
+                                 f"Your vertices have shape {vertices.shape} "
+                                 f"but your codes have shape {codes.shape}")
             if len(codes) and codes[0] != self.MOVETO:
                 raise ValueError("The first element of 'code' must be equal "
-                                 "to 'MOVETO' ({})".format(self.MOVETO))
+                                 f"to 'MOVETO' ({self.MOVETO}).  "
+                                 f"Your first code if {codes[0]}")
```
Comment (QuLogic) [inline lib/matplotlib/path.py]:
```suggestion
                                 f"Your first code is {codes[0]}")
```

File: lib/matplotlib/tests/test_path.py [hunk 6]
```diff
@@ -1,5 +1,5 @@
 import copy
-
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_path.py]: Restore blank line after `re`.

File: lib/matplotlib/cbook/__init__.py [hunk 7]
```diff
@@ -2194,6 +2194,44 @@ def _check_in_list(_values, **kwargs):
...
+        if len(target_shape) != len(data_shape) or any(
+                t not in [s, None]
+                for t, s in zip(target_shape, data_shape)
+        ):
+            dim_labels = iter(itertools.chain(
+                'MNLIJKLH',
+                (f"D{i}" for i in itertools.count())))
+            text_shape = ", ".join(str(_) for _ in
+                                   (n if n is not None else next(dim_labels)
+                                    for n in target_shape))
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py]: I don't understand why you have two `for`s; isn't this just:
```suggestion
            text_shape = ", ".join(str(n) if n is not None
                                   else next(dim_labels)
                                   for n in target_shape)
```

File: lib/matplotlib/cbook/__init__.py [hunk 8]
```diff
@@ -2194,6 +2194,44 @@ def _check_in_list(_values, **kwargs):
                 .format(v, k, ', '.join(map(repr, values))))
 
 
+def _check_shape(_shape, **kwargs):
+    """
+    For each *key, value* pair in *kwargs*, check that *value* has the shape
+    *_shape*, if not, raise an appropriate ValueError.
+
+    None in the shape is treated as a "free" sizes that can have any length.
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py]:
```suggestion
    *None* in the shape is treated as a "free" size that can have any length.
```

File: lib/matplotlib/cbook/__init__.py [hunk 9]
```diff
@@ -2194,6 +2194,44 @@ def _check_in_list(_values, **kwargs):
...
+    e.g. (None, 2) -> (N, 2)
+
+    The values checked must be numpy arrays.
+
+    Examples
+    --------
+    To check from (N, 2) shaped arrays
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py]:
```suggestion
    To check for (N, 2) shaped arrays
```

File: lib/matplotlib/cbook/__init__.py (hunk 7 above)
Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py]: Yup, I was making it more complicated than it needed to be.

**Pull request merged**